scipy/ndimage/_measurements.py

```diff
@@ -307,6 +307,115 @@ def find_objects(input, max_label=0):
     return _nd_image.find_objects(input, max_label)
 
 
+def value_indices(arr, *, ignore_value=None):
+    """
+    Find indices of each distinct value in given array.
+
+    Parameters
+    ----------
+    arr : ndarray of ints
+        Array containing integer values.
+    ignore_value : int, optional
+        This value will be ignored in searching the `arr` array. If not
+        given, all values found will be included in output. Default
+        is None.
+
+    Returns
+    -------
+    indices : dictionary
+        A Python dictionary of array indices for each distinct value. The
+        dictionary is keyed by the distinct values, the entries are array
+        index tuples covering all occurrences of the value within the
+        array.
+
+        This dictionary can occupy significant memory, usually several times
+        the size of the input array.
+
+    Notes
+    -----
+    For a small array with few distinct values, one might use
+    `numpy.unique()` to find all possible values, and ``(arr == val)`` to
+    locate each value within that array. However, for large arrays,
+    with many distinct values, this can become extremely inefficient,
+    as locating each value would require a new search through the entire
+    array. Using this function, there is essentially one search, with
+    the indices saved for all distinct values.
+
+    This is useful when matching a categorical image (e.g. a segmentation
+    or classification) to an associated image of other data, allowing
+    any per-class statistic(s) to then be calculated. Provides a
+    more flexible alternative to functions like ``scipy.ndimage.mean()``
+    and ``scipy.ndimage.variance()``.
+
+    Note for IDL users: this provides functionality equivalent to IDL's
```

Comment:
Thanks @stefanv, yes, I agree. 

I have added a ref to the scikit-image function, and I also realised that scipy.stats.binned_statistic was potentially relevant, so I added that, too. 

All of which revives my worries that I am just adding to fragmentation of the Python ecosystem, but I think/hope that this model of returning the indices provides a good general solution to a range of needs. 

On the IDL ref - the inspiration for my very first version of this was one of my users asking for a Python equivalent of what they used to do in IDL. More recently, an Internet search of these terms also finds a similar question on StackOverflow, so I thought it might help others discover it.

Reply:
P.S. In general, discoverability is a growing problem, and this is why I think it is important to include these kind of cross references.